File: src/adapters/provider.ts
import type { IrRequest } from '../ir/types.js';

export type ProviderRequestBody = Record<string, unknown>;

/** The denormalized request, as it would go on the wire. */
export type UpstreamCall = {
  body: string;
  /** Forwardable headers: Caveman's own have already been stripped. */
  headers: Headers;
  /** The client's headers as they arrived, for adapter-specific Caveman options. */
  originalHeaders: Headers;
  signal: AbortSignal;
};

/**
 * How a provider reaches its upstream. Exists because not every provider is an
 * HTTP host: a local CLI is a subprocess, and only the adapter knows how to
 * speak to it. The result is a `Response` either way, so the handler treats
 * both alike.
 */
export type ProviderTransport = (call: UpstreamCall) => Promise<Response>;

/**
 * Everything the HTTP layer needs to serve one provider. A provider owns its
 * route, its wire format, and the shape of its errors, so adding one is adding
 * an implementation of this type rather than editing the handler.
 */
export type ProviderAdapter = {
  name: string;
  /** Route this provider serves, and the upstream path requests forward to. */
  path: string;
  /** Where this provider's requests are forwarded, absent an env override. */
  baseUrl: string;
  toIR(body: ProviderRequestBody): IrRequest;
  fromIR(request: IrRequest): ProviderRequestBody;
  /** Wraps a Caveman-generated message in the provider's own error shape. */
  errorEnvelope(message: string): ProviderRequestBody;
  /** Sends the request upstream. Absent means an HTTP POST to `baseUrl + path`. */
  transport?: ProviderTransport;
};

File: internal/adapters/provider.go
package adapters

import "github.com/carlelieser/caveman/internal/ir"

// RequestBody is one provider's wire body, in the key order it arrived in.
type RequestBody = *ir.Object

// Provider is everything the HTTP layer needs to serve one provider. A provider
// owns its route, its wire format, and the shape of its errors, so adding one
// is adding an implementation of this interface rather than editing the
// handler.
type Provider interface {
	Name() string
	// Path is the route this provider serves, and the upstream path requests
	// forward to.
	Path() string
	// BaseURL is where this provider's requests are forwarded, absent an env
	// override.
	BaseURL() string
	ToIR(body RequestBody) ir.Request
	FromIR(request ir.Request) RequestBody
	// ErrorEnvelope wraps a Caveman-generated message in the provider's own
	// error shape.
	ErrorEnvelope(message string) RequestBody
}

// Registry is the set of providers the HTTP layer serves. Registering a second
// provider is appending to this slice; no handler knows any provider by name.
type Registry []Provider

// ByPath finds the provider that claims a route. A route no provider claims is
// not served.
func (r Registry) ByPath(path string) (Provider, bool) {
	for _, provider := range r {
		if provider.Path() == path {
			return provider, true
		}
	}
	return nil, false
}
